### PR TITLE
Add support for creating confidential assets.

### PR DESCRIPTION
**Pull request by ParnianAlimi**
Addresses CRYP-153. Adds an extrinsic to register a confidential asset.

### CODE REVIEW

**Review by aafshar-poly — APPROVED** [review]

File: pallets/common/src/traits/asset.rs
Comment (Centril) [inline pallets/common/src/traits/asset.rs:71]: Could you give this a doc comment? (Use `fn create_asset` as a guide.)

File: pallets/confidential-asset/src/lib.rs
Comment (Centril) [inline pallets/confidential-asset/src/lib.rs:92]:
```suggestion
```
Doesn't really add anything.

Comment (Centril) [reply pallets/confidential-asset/src/lib.rs:92]: Also, storage migration?

File: pallets/confidential-asset/src/lib.rs
Comment (Centril) [inline pallets/confidential-asset/src/lib.rs:178]:
```suggestion
            <ConfidentialTickers>::append(AssetId { id: ticker.as_bytes().clone() });
```

File: pallets/runtime/tests/src/settlement_test.rs
Comment (Centril) [inline pallets/runtime/tests/src/settlement_test.rs:10]:
```suggestion
use pallet_asset as asset;
```

File: pallets/runtime/tests/src/ext_builder.rs
Comment (Centril) [inline pallets/runtime/tests/src/ext_builder.rs:3]:
```suggestion
use pallet_asset as asset;
```

File: pallets/runtime/tests/src/dividend_test.rs
Comment (Centril) [inline pallets/runtime/tests/src/dividend_test.rs:10]:
```suggestion
use pallet_asset as asset;
```

File: pallets/runtime/tests/src/confidential_test.rs
Comment (Centril) [inline pallets/runtime/tests/src/confidential_test.rs:7]:
```suggestion
use pallet_asset as asset;
```

File: pallets/runtime/tests/src/voting_test.rs
Comment (Centril) [inline pallets/runtime/tests/src/voting_test.rs:5]:
```suggestion
use pallet_asset as asset;
```

File: pallets/runtime/tests/src/statistics_test.rs
Comment (Centril) [inline pallets/runtime/tests/src/statistics_test.rs:5]:
```suggestion
use pallet_asset as asset;
```

File: pallets/runtime/tests/src/confidential_asset_test.rs
Comment (Centril) [inline pallets/runtime/tests/src/confidential_asset_test.rs:58]:
```suggestion
        let funding_round_name = b"round1".into();
```
Shouldn't be necessary, as `create_confidential_asset` should constrain the type.

File: pallets/runtime/tests/src/confidential_asset_test.rs
Comment (Centril) [inline pallets/runtime/tests/src/confidential_asset_test.rs:216]:
```suggestion
        let valid_asset_ids = ConfidentialAsset::confidential_tickers();
```

File: pallets/runtime/tests/src/confidential_asset_test.rs
Comment (Centril) [inline pallets/runtime/tests/src/confidential_asset_test.rs:126]: Please use byte-string notation instead for improved readability.

File: pallets/common/src/traits/asset.rs
Comment (Centril) [inline pallets/common/src/traits/asset.rs:71]: Are there ever going to be more than one implementation of this?
Otherwise, are you doing this to break a circular dependency?
If we can avoid unnecessary abstractions generics, then that's good.

Comment (ParnianAlimi) [reply pallets/common/src/traits/asset.rs:71]: Yes, we are doing this to avoid a circular dependency, these traits are mostly used to call a functionality of a pallet from another pallet, in this case we are keeping confidential asset's logic separate from regular asset logic, but because of some business requirements we still need to use a single storage (placed in the asset pallet) to make sure the asset wasn't registered before, etc.

File: pallets/confidential-asset/src/lib.rs
Comment (ParnianAlimi) [reply pallets/confidential-asset/src/lib.rs:92]: Other storage API in other pallets are documented like this, e.g.:
```
        /// Ticker registration details.
        /// (ticker) -> TickerRegistration
        pub Tickers get(fn ticker_registration): map hasher(blake2_128_concat) Ticker => TickerRegistration<T::Moment>;
```
by `() -> List of confidential tickers` I mean this getter doesn't take an input, and returns a list of assets.

what do you mean by storage migration?

File: pallets/runtime/tests/src/confidential_asset_test.rs
Comment (ParnianAlimi) [reply pallets/runtime/tests/src/confidential_asset_test.rs:58]: apparently `into()` can't infer the type, unless I make the it explicit.

File: pallets/common/src/traits/asset.rs
Comment (Centril) [reply pallets/common/src/traits/asset.rs:72]: Btw, I wouldn't say its a good thing that these trait functions have no documentation.

File: pallets/confidential-asset/src/lib.rs
Comment (Centril) [reply pallets/confidential-asset/src/lib.rs:92]: - Those storage items are typically maps, where details like the hasher, etc. obscure semantics, but this is just a vector, so I still don't think the comment adds anything.

- Storage migration refers, in this case, to writing code that renames the old storage item into the new one, by reading the storage item, removing the old one, and then putting it under the new storage item name.

File: pallets/runtime/tests/src/confidential_asset_test.rs
Comment (Centril) [reply pallets/runtime/tests/src/confidential_asset_test.rs:58]: Ah, that's probably because the method dispatch is forcing "we must know the type at this point" since rustc is using bidirectional type-checking instead of more constraint-solving-based unification.

File: pallets/common/src/traits/asset.rs
Comment (Centril) [reply pallets/common/src/traits/asset.rs:72]: Hmm; but neither the confidential crate nor the asset crate refer to each other at the moment, so the confidential one could ostensibly add the asset crate as a dependency... I guess I'm not seeing where the circular dependency would arise from. Can you elaborate on that?

File: pallets/confidential-asset/src/lib.rs
Comment (satyamakgec) [inline pallets/confidential-asset/src/lib.rs:122]: Is anyone allowed to call this dispatchable?

File: pallets/confidential-asset/src/lib.rs
Comment (vkomenda) [inline pallets/confidential-asset/src/lib.rs:81]: These extra 2 spaces of indentation could be removed.

File: pallets/confidential-asset/src/lib.rs
Comment (vkomenda) [inline pallets/confidential-asset/src/lib.rs:87]: Extra 2 spaces.

File: pallets/confidential-asset/src/lib.rs
Comment (vkomenda) [inline pallets/confidential-asset/src/lib.rs:93]: Any access would require decoding the entire `Vec`. There is a way around that, just use a characteristic function instead - a map from `AssetId` to `bool`.

`BTreeSet` would be better than `Vec`: `log(n)` reads as opposed to `n`. But again it would require reading the whole `BTreeSet` from runtime storage, which can be avoided.

**Review by vkomenda — COMMENTED** [review]
LGTM, just a few suggestions.

File: pallets/confidential-asset/src/lib.rs
Comment (vkomenda) [inline pallets/confidential-asset/src/lib.rs:80]: I wonder if there is a typo. Should the fn name be `mercat_accounts`?

File: pallets/confidential-asset/src/lib.rs
Comment (satyamakgec) [reply pallets/confidential-asset/src/lib.rs:93]: @aafshar-poly & I had a discussion about this couple of weeks ago. Arash do you guys still need the whole vector of asset_ids to validate the proof? I think it was more related to the index of asset_id. we can re-discuss this.

File: pallets/common/src/traits/asset.rs
Comment (ParnianAlimi) [reply pallets/common/src/traits/asset.rs:72]: to create a confidential asset I need to access the storage in both `asset` and `confidential-asset` pallets, confidential asset pallet already depends on the asset pallet, if I don't break out the logic like I did, and I put `create_confidential_asset()` in the asset pallet, I'll have to make `asset` depend on the confidential pallet, and that's a circular dependency.

File: pallets/confidential-asset/src/lib.rs
Comment (ParnianAlimi) [reply pallets/confidential-asset/src/lib.rs:122]: do you mean the `confidential_tickers()`? yes, wallet will need to call it eventually.

File: pallets/common/src/traits/asset.rs
Comment (ParnianAlimi) [reply pallets/common/src/traits/asset.rs:72]: If anything, I'm all for documentation when it adds value, it's just that I'm not a PolyMesh developer and I don't fully understand what the rest of these trait functions are doing, I just didn't want to add a full comment section for this one and leave the rest undocumented.

File: pallets/confidential-asset/src/lib.rs
Comment (satyamakgec) [reply pallets/confidential-asset/src/lib.rs:122]: Sorry, I had to specific. I mean the dispatchable `validate_mercat_account()`. Let's say I am a polymesh user and I want to invest in some confidential asset then I will call this function or someone on my behalf?

File: pallets/asset/src/lib.rs
Comment (ParnianAlimi) [reply pallets/asset/src/lib.rs:1961]: Yeah, when I wrote this change we were still thinking of adding the minting process to asset creation. I'll exclude them.

File: pallets/confidential-asset/src/lib.rs
Comment (ParnianAlimi) [reply pallets/confidential-asset/src/lib.rs:122]: Yes, before your MERCAT account gets added someone has to validate that you are in fact planning to hold an asset that's among confidential assets, and you start from a 0 balance. Since the asset type is hidden from MESH the user will have to call this function to show their account is valid.

File: pallets/asset/src/lib.rs
Comment (ParnianAlimi) [reply pallets/asset/src/lib.rs:1943]: Once Arash's code goes in, we will be calling this function with `total_supply = 0`, so it won't really do anything meaningful, but I'll make it more explicit.

File: pallets/confidential-asset/src/lib.rs
Comment (ParnianAlimi) [reply pallets/confidential-asset/src/lib.rs:80]: Good catch.

File: pallets/confidential-asset/src/lib.rs
Comment (ParnianAlimi) [reply pallets/confidential-asset/src/lib.rs:93]: Yes, we need the whole AssetID vector. The list of all confidential asset IDs ever created to be exact.

File: pallets/asset/src/lib.rs
Comment (ParnianAlimi) [reply pallets/asset/src/lib.rs:1943]: Thinking more about this, are you sure we don't need to update stats here? we are still adding a primary issuer for the asset and that should increment the number of investors of the asset.

File: pallets/confidential-asset/src/lib.rs
Comment (vkomenda) [inline pallets/confidential-asset/src/lib.rs:47]: In general it would be good to have doc comments both on type parameters and the trait itself.

File: pallets/runtime/tests/src/confidential_asset_test.rs
Comment (vkomenda) [inline pallets/runtime/tests/src/confidential_asset_test.rs:216]: Unused variable.

File: pallets/confidential-asset/src/lib.rs
Comment (vkomenda) [reply pallets/confidential-asset/src/lib.rs:93]: OK, got you. The only place I found in in this PR where `confidential_tickers` is called is an application of `AccountCreator::create`. It takes a reference to the `Vec`. I would suggest modifying `AccountCreator::create` to take `&[Scalar]` instead for uniformity and also because slices are better as arguments as opposed to `&Vec<_>`.

**Review by vkomenda — APPROVED** [review]
LGTM, a few suggestions.

File: pallets/asset/src/lib.rs
Comment (satyamakgec) [reply pallets/asset/src/lib.rs:1942]: We are not really assigning the balance to the primary issuer so it doesn't make sense to call this.

Comment (satyamakgec) [reply pallets/asset/src/lib.rs:1942]: But we can capture this in another story if you want without delaying this PR approval.

File: pallets/asset/src/lib.rs
Comment (satyamakgec) [inline pallets/asset/src/lib.rs:1962]: Realistically we can cover this in the above `if` statement

**Review by satyamakgec — APPROVED** [review]
Some minor nit-picks if you want to taker care before the merge 😉

**Pull request merged**
